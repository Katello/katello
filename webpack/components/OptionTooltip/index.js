import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import { Popover, OverlayTrigger } from 'patternfly-react';
import './OptionTooltip.scss';

class OptionTooltip extends Component {
  constructor(props) {
    super(props);

    this.state = {
      tooltipOpen: false,
    };
    this.handleInputChange = this.handleInputChange.bind(this);
    this.renderTooltip = this.renderTooltip.bind(this);
  }

  handleInputChange(event, index) {
    const { options } = this.props;
    options[index].value = event.target.checked;
    this.setState(options);
    this.props.onChange(options);
  }
  renderTooltip() {
    const { options, id } = this.props;
    return (
      <Popover id={id} className="option-tooltip">
        <ul>
          {
            options.map((option, index) => (
              <li key={option.key}>
                <input type="checkbox" checked={option.value} name={option.key} id={option.key} onChange={e => this.handleInputChange(e, index)} />
                <span>{option.label}</span>
              </li>
            ))
          }
        </ul>
      </Popover>
    );
  }
  render() {
    const { icon, options, rootClose } = this.props;
    const onOpen = () => {
      this.setState({ tooltipOpen: true });
    };
    const onClose = () => {
      this.props.onClose(options);
      this.setState({ tooltipOpen: false });
    };

    return (
      <OverlayTrigger
        overlay={this.renderTooltip()}
        placement="bottom"
        trigger={['click']}
        rootClose={rootClose}
        onEnter={onOpen}
        onExit={onClose}
      >
        <i className={classNames('fa', icon, 'tooltip-button', { 'tooltip-open': this.state.tooltipOpen })} />
      </OverlayTrigger>
    );
  }
}

OptionTooltip.propTypes = {
  icon: PropTypes.string.isRequired,
  id: PropTypes.string.isRequired,
  options: PropTypes.arrayOf(PropTypes.shape({
    key: PropTypes.string,
    label: PropTypes.string,
    value: PropTypes.bool,
  })).isRequired,
  onChange: PropTypes.func,
  onClose: PropTypes.func,
  rootClose: PropTypes.bool,
};
OptionTooltip.defaultProps = {
  onChange: () => {},
  onClose: () => {},
  rootClose: true,
};
export default OptionTooltip;
