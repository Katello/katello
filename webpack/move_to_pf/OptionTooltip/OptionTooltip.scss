@import '~patternfly/dist/sass/patternfly/variables';
.option-tooltip {
  ul {
    list-style-type: none;
    padding: 0;
  }
  .popover-content{
    min-width: 120px;
    max-width: 160px;
  }
  span {
    margin-left: 8px;
  }
}
.tooltip-open {
  color: $color-pf-blue-400;
}
.tooltip-button {
  cursor: pointer;
}
