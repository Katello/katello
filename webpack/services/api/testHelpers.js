
export const failureAction = (type, message = 'Request failed with status code 422') => (
  {
    type,
    payload: {
      messages: [message],
      result: new Error(message),
    },
  }
);

export const toastErrorAction = (message = 'Request failed with status code 422') => (
  {
    payload: {
      message: {
        message,
        sticky: true,
        type: 'error',
      },
    },
    type: 'TOASTS_ADD',
  }
);

export default {
  failureAction,
  toastErrorAction,
};
