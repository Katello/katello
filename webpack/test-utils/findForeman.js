const path = require('path');
const fs = require('fs');

const getForemanLocation = () => {
  const foremanLocations = ['./foreman', '../foreman', '../../foreman'];
  const notFound = 'Foreman directory cannot be found! This action requires Foreman to be present ' +
  'in either a parent, sibling, or child directory relative to the plugin.';
  const currentDir = process.cwd();
  let fullPath;

  foremanLocations.forEach((relativeForemanPath) => {
    const possibleForemanPath = path.join(currentDir, relativeForemanPath);
    if (fs.existsSync(possibleForemanPath)) fullPath = possibleForemanPath;
  });

  if (!fullPath) throw new Error(notFound);
  return fullPath;
};

// Get a subdirectory within Foreman
const getForemanRelativePath = (relativeForemanPath) => {
  const foremanLocation = getForemanLocation();
  const notFound = `Could not find ${relativeForemanPath} in ${foremanLocation}`;
  const foremanRelativePath = path.join(foremanLocation, relativeForemanPath);
  if (!fs.existsSync(foremanRelativePath)) throw new Error(notFound);
  return foremanRelativePath;
};

module.exports = { getForemanLocation, getForemanRelativePath };
