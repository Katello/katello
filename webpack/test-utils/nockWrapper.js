import nock from 'nock';

// Using the library 'nock' as it matches actual network requests rather than mock another
// library. This is helpful when the request is not coming from Katello. For example, axios
// called within Katello can be mocked with axios-mock-adapter or similar, but a http request
// made by axios that is coming from Foreman cannot be mocked by axios-mock-adapter or a
// jest mock within Katello. So to do this, we can mock the request a level deeper within
// nodejs by using nock.
export const nockInstance = nock('http://localhost');

// Calling .done() with nock asserts that the request was fufilled. We use a timeout to ensure
// that the component has set up and made the request before the assertion is made.
export const assertNockRequest = (scope, timeout = 10000) => {
  setTimeout(() => {
    scope.done();
  }, timeout);
};
