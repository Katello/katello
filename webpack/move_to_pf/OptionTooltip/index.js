import React, { Component } from '@theforeman/vendor/react';
import PropTypes from '@theforeman/vendor/prop-types';
import classNames from '@theforeman/vendor/classnames';
import { Popover, OverlayTrigger } from '@theforeman/vendor/patternfly-react';
import './OptionTooltip.scss';

class OptionTooltip extends Component {
  constructor(props) {
    super(props);

    this.state = {
      tooltipOpen: false,
    };
    this.handleInputChange = this.handleInputChange.bind(this);
    this.renderTooltip = this.renderTooltip.bind(this);
  }

  handleInputChange(event, index) {
    const { options } = this.props;
    options[index].value = event.target.checked;
    this.setState(options);
    this.props.onChange(options);
  }
  renderTooltip() {
    const { options, id } = this.props;
    return (
      <Popover id={id} className="option-tooltip">
        <ul>
          {
            options.map((option, index) => (
              <li key={option.key}>
                <input type="checkbox" checked={option.value} name={option.key} id={option.key} onChange={e => this.handleInputChange(e, index)} />
                <span>{option.label}</span>
              </li>
            ))
          }
        </ul>
      </Popover>
    );
  }
  render() {
    const { icon, options, rootClose } = this.props;
    const onOpen = () => {
      this.setState({ tooltipOpen: true });
    };
    const onClose = () => {
      this.props.onClose(options);
      this.setState({ tooltipOpen: false });
    };

    return (
      <OverlayTrigger
        overlay={this.renderTooltip()}
        placement="bottom"
        trigger={['click']}
        rootClose={rootClose}
        onEnter={onOpen}
        onExit={onClose}
      >
        <i className={classNames('fa', icon, 'tooltip-button', { 'tooltip-open': this.state.tooltipOpen })} />
      </OverlayTrigger>
    );
  }
}

OptionTooltip.propTypes = {
  icon: PropTypes.string.isRequired,
  id: PropTypes.string.isRequired,
  options: PropTypes.arrayOf(PropTypes.shape({
    key: PropTypes.string,
    label: PropTypes.string,
    value: PropTypes.bool,
  })).isRequired,
  onChange: PropTypes.func,
  onClose: PropTypes.func,
  rootClose: PropTypes.bool,
};
OptionTooltip.defaultProps = {
  onChange: () => {},
  onClose: () => {},
  rootClose: true,
};
export default OptionTooltip;
