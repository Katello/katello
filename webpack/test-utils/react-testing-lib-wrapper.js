// For react-testing-library helpers, overrides, and utilities
// All elements from react-testing-library can be imported from this wrapper.
// See https://testing-library.com/docs/react-testing-library/setup for more info
import React from 'react';
import thunk from 'redux-thunk';
import Immutable from 'seamless-immutable';
import { APIMiddleware, reducers as apiReducer } from 'foremanReact/redux/API';
import { reducers as fillReducers } from 'foremanReact/components/common/Fill';
import { reducers as foremanModalReducer } from 'foremanReact/components/ForemanModal';
import { STATUS } from 'foremanReact/constants';
import { render, waitFor, waitForElementToBeRemoved } from '@testing-library/react';
import { createStore, applyMiddleware, combineReducers } from 'redux';
import { Provider } from 'react-redux';
import { MemoryRouter, BrowserRouter } from 'react-router-dom';
import { initialSettingsState } from '../scenes/Settings/SettingsReducer';
import allKatelloReducers from '../redux/reducers/index.js';

// r-t-lib's print limit for debug() is quite small, setting it to a much higher char max here.
// See https://github.com/testing-library/react-testing-library/issues/503 for more info.
process.env.DEBUG_PRINT_LIMIT = 99999;

// Renders testable component with redux and react-router according to Katello's usage
// This should be used when you want a fully connected component with Redux state and actions.
function renderWithRedux(
  component,
  {
    apiNamespace, // namespace if using API middleware
    initialApiState = { response: {}, status: STATUS.PENDING }, // Default state for API middleware
    initialState = {}, // Override full state
    routerParams = {},
  } = {},
) {
  // Adding the reducer in the expected namespaced format
  const combinedReducers = combineReducers({
    katello: allKatelloReducers,
    ...apiReducer,
    ...foremanModalReducer,
    ...fillReducers,
  });

  // Namespacing the initial state as well
  const initialFullState = Immutable({
    API: {
      [apiNamespace]: initialApiState,
    },
    katello: {
      settings: {
        settings: initialSettingsState,
      },
    },
    extendable: {},
    ...initialState,
  });
  const middlewares = applyMiddleware(thunk, APIMiddleware);
  const store = createStore(combinedReducers, initialFullState, middlewares);
  const connectedComponent = (
    <Provider store={store}>
      <MemoryRouter {...routerParams} >{component}</MemoryRouter>
    </Provider>
  );

  return { ...render(connectedComponent), store };
}

// When you actually need to change browser history
const renderWithRouter = (ui, { route = '/' } = {}) => {
  window.history.pushState({}, 'Test page', route);

  return render(ui, { wrapper: BrowserRouter });
};

// When the tests run slower, they can hit the default waitFor timeout, which is 1000ms
// There doesn't seem to be a way to set it globally for r-t-lib, so using this wrapper function
const rtlTimeout = 5000;
export const patientlyWaitFor = waitForFunc => waitFor(waitForFunc, { timeout: rtlTimeout });
export const patientlyWaitForRemoval = waitForFunc =>
  waitForElementToBeRemoved(waitForFunc, { timeout: rtlTimeout });

// re-export everything, so the library can be used from this wrapper.
export * from '@testing-library/react';

export { renderWithRedux, renderWithRouter };
