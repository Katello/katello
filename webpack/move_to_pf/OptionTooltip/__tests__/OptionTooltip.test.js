import OptionTooltip from '../';
import { testComponentSnapshotsWithFixtures } from '../../test-utils/testHelpers';

const onClose = () => {};
const options = [
  {
    key: 'option1',
    value: true,
    label: 'One',
  },
  {
    key: 'option2',
    value: false,
    label: 'Two',
  },
];

const fixtures = {
  'renders ': {
    icon: 'test',
    id: 'test',
    options: [],
    onClose,
  },
  'renders a list of options with default values ': {
    icon: 'test',
    id: 'test',
    options,
    onClose,
  },
};
describe('OptionTooltip', () => testComponentSnapshotsWithFixtures(OptionTooltip, fixtures));
