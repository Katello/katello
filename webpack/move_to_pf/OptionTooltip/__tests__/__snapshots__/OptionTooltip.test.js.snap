// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OptionTooltip renders  1`] = `
<OverlayTrigger
  defaultOverlayShown={false}
  onEnter={[Function]}
  onExit={[Function]}
  overlay={
    <Popover
      bsClass="popover"
      className="option-tooltip"
      id="test"
      placement="right"
    >
      <ul />
    </Popover>
  }
  placement="bottom"
  rootClose={true}
  trigger={
    Array [
      "click",
    ]
  }
>
  <i
    className="fa test tooltip-button"
  />
</OverlayTrigger>
`;

exports[`OptionTooltip renders a list of options with default values  1`] = `
<OverlayTrigger
  defaultOverlayShown={false}
  onEnter={[Function]}
  onExit={[Function]}
  overlay={
    <Popover
      bsClass="popover"
      className="option-tooltip"
      id="test"
      placement="right"
    >
      <ul>
        <li>
          <input
            checked={true}
            id="option1"
            name="option1"
            onChange={[Function]}
            type="checkbox"
          />
          <span>
            One
          </span>
        </li>
        <li>
          <input
            checked={false}
            id="option2"
            name="option2"
            onChange={[Function]}
            type="checkbox"
          />
          <span>
            Two
          </span>
        </li>
      </ul>
    </Popover>
  }
  placement="bottom"
  rootClose={true}
  trigger={
    Array [
      "click",
    ]
  }
>
  <i
    className="fa test tooltip-button"
  />
</OverlayTrigger>
`;
